example_code/shmem_ctx_pipelined_reduce.c

```diff
@@ -0,0 +1,56 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <shmem.h>
+
+#define LEN  8192 /* Full buffer length */
+#define PLEN 512  /* Length of each pipeline stage */
+
+int in_buf[LEN], out_buf[LEN];
+
+int main(void) {
+    int i, j, *pbuf[2];
+    shmem_ctx_t ctx[2];
+
+    shmem_init();
+    int me = shmem_my_pe();
+    int npes = shmem_n_pes();
+
+    pbuf[0] = shmem_malloc(PLEN * npes * sizeof(int));
+    pbuf[1] = shmem_malloc(PLEN * npes * sizeof(int));
+
+    int ret_0 = shmem_ctx_create(0, &ctx[0]);
+    int ret_1 = shmem_ctx_create(0, &ctx[1]);
```

Comment:
I do not think so: If the contexts are private, the barrier would have no effect, on either of them. 

Now, that has raised my attention to the fact that barrier_all has an effect on all shared contexts. 

The original intent of PRIVATE/SHARED was to pertain to thread locality (i.e. an operation on a private context does not have to care for making it consistent with possible concurrent access from other threads). I think the conflation of two different concepts here is a detriment with PRIVATE/SHARED being confounded with DEFAULT/Non default. 

I was also discussing during the call yesterday that we could have an "IN: ctx" as the last optional param to shmem_barrier[_all]; Jim had a good comeback that in many instances you may want to quiet a couple of ctx before doing the barrier, which would make the usage asymmetric, which is kinda ugly.
```
shmem_ctx_quiet(ctx1);
shmem_ctx_quiet(ctx2);
shmem_barrier_all(ctx3);
```